### PR TITLE
Quick start guide for CAPBM

### PR DESCRIPTION
**Pull request by Xenwar**
This commit introduces the architecture of CAPBM in that:
- It shows components (controllers and CRs) involved
- Flow of information between CRS

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -24,6 +24,9 @@ API used with this `cluster-api` provider. You can also see the [cluster
 deployment workflow](docs/deployment_workflow.md) for the outline of the
 deployment process.
 
+## Architecture
+The Architecture with the components involved is documented [here](docs/architecture.md)
```
Comment (kashifest) [inline README.md]: architecture

File: docs/architecture.md [hunk 2]
```diff
@@ -0,0 +1,589 @@
+# Architecture
+
+## Introduction
+
+The ```cluster-api-provider-baremetal (CAPBM)``` one of the controllers involved in managing the life cycle of kubernetes clusters on Bare Metal Machines. This document describes the components involved and their roles. It also discusses the flow of information from one CR to another with the help of the controllers.
```
Comment (kashifest) [inline docs/architecture.md]: is one of the controllers

File: docs/architecture.md [hunk 3]
```diff
@@ -0,0 +1,589 @@
...
+
+
+## Components
+The ```cluster-api-provider-baremetal (CAPBM)``` operator is responsible watching and reconciling multiple resources. However, it is important to see other controllers and custom resources (CRs) involved in the process. The ultimate goal of the interaction between the controllers and CRs is to provision a kubernetes cluster on Bare Metal Machines. To that end, the controllers perform different actions on one or more relevant CRs.
```
Comment (kashifest) [inline docs/architecture.md]: responsible for watching

File: docs/architecture.md [hunk 4]
```diff
@@ -0,0 +1,589 @@
...
+
+The following diagram shows the different controllers and CRs involved. The CAPI, CABPK and CAPBM controllers are beyond the scope of this document. With respect to CAPBM, we focus on what CRS it `watches` and `reconciles`. The arrows in <span style="color:black">**black**</span> show which CRs the controller `reconciles` while the one in <span style="color:red">**red**</span> show a `watching` relationship.
```
Comment (kashifest) [inline docs/architecture.md]: watching relationship sounds a bit weird????

File: docs/architecture.md [hunk 5]
```diff
@@ -0,0 +1,589 @@
...
+
+![components](images/components.png)
+
+As shown in the above Components' diagram, CAPBM is watching for changes in `Machine` CR and upon change, it makes changes on the `BareMetalMachine` CR. Similarly, it watches a `Cluster` CR and makes changes on a related `BareMetalCluster` CR.
```
Comment (kashifest) [inline docs/architecture.md]: a cluster CR of cluster CR

File: docs/architecture.md [hunk 6]
```diff
@@ -0,0 +1,589 @@
...
+
+The left most components, BMO controller and BareMetalHost(BMH) CR, are the closest to the Bare Metal Machine. If one wants to changes the state of a Bare Metal Machine, they modify the BMH CR. Upon change to the BMH, BMO interacts with Ironic to make changes on the Bare Metal Machine.
+
+The above logic works in the opposite direction as well. If there is any change on the Bare Meta machine, BMO learns about the change(s) via Ironic and a chain of events starts. BMO, after learning about the changes, makes the required changes on the BMH CR and CAPBM, upon seeing the changes on a BMH resource, it makes the relevant changes on the BareMetalMachine CR and so on.
```
Comment (kashifest) [inline docs/architecture.md]: via ironic

Comment (kashifest) [reply docs/architecture.md]: Once BMO learns about the changes, it makes the required changes ...

File: docs/architecture.md (hunk 5 above)
Comment (Xenwar) [reply docs/architecture.md]: I need more info on this one.

**Comment by Xenwar** [comment]
test-v1a2-integration

File: docs/architecture.md [hunk 7]
```diff
@@ -0,0 +1,456 @@
...
+The left most components, BMO controller and BareMetalHost(BMH) CR, are the
+closest to the Bare Metal Machine. If one wants to changes the state of a Bare
+Metal Machine, they modify the BMH CR. Upon change to the BMH, BMO interacts
+with Ironic to make changes on the Bare Metal Machine.
+
+The above logic works in the opposite direction as well. If there is any change
+on the Bare Meta machine, BMO learns about the change(s) via ironic and a chain
+of events starts. Once BMO learns about the changes, it makes the required
+changes on the BMH CR and CAPBM, upon seeing the changes on a BMH resource, it
+makes the relevant changes on the BareMetalMachine CR and so on.
```
Comment (maelk) [inline docs/architecture.md]: This is not really correct. At the moment there is nothing in place to return infor from BMH to BMM.

File: docs/architecture.md (hunk 3 above)
Comment (Xenwar) [reply docs/architecture.md]: done

File: docs/architecture.md (hunk 4 above)
Comment (Xenwar) [reply docs/architecture.md]: done

File: docs/architecture.md (hunk 5 above)
Comment (Xenwar) [reply docs/architecture.md]: done

File: docs/architecture.md (hunk 6 above)
Comment (Xenwar) [reply docs/architecture.md]: done

File: docs/architecture.md (hunk 7 above)
Comment (Xenwar) [reply docs/architecture.md]: done

File: docs/architecture.md [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+Similarly, it watches `Cluster` CR and makes changes on a related
+`BareMetalCluster` CR.
+
+The left most components, BMO controller and BareMetalHost(BMH) CR, are the
+closest to the Bare Metal Machine. If one wants to changes the state of a Bare
+Metal Machine, they modify the BMH CR. Upon change to the BMH, BMO interacts
+with Ironic to make changes on the Bare Metal Machine.
+
+The above logic works in the opposite direction as well. If there is any change
+on the Bare Meta machine, BMO learns about the change(s) via ironic and a chain
```
Comment (maelk) [inline docs/architecture.md]: Metal

File: docs/architecture.md [hunk 9]
```diff
@@ -0,0 +1,455 @@
...
+of events starts. Once BMO learns about the changes, it makes the required
+changes on the BMH.
```
Comment (maelk) [inline docs/architecture.md]: Only true for the initial introspection and state changes

File: docs/architecture.md [hunk 10]
```diff
@@ -0,0 +1,455 @@
...
+#### Flow of information
+
+As was shown on the above CRs, some of the fields are introduced in one CR and
+they travel through multiple CRs to reach the BMH, which is the closest to the
+Bare Metal Machine. There is also a movement of information from the
+virtual/physical machines towards the CRs, but this is beyond the scope of this
+ document. Some of the fields are added by users, while the others are by the
+ relevant controllers.
+
+We have add the source of relevant fields as comments in the above yaml files.
```
Comment (maelk) [inline docs/architecture.md]: added

File: docs/architecture.md [hunk 11]
```diff
@@ -0,0 +1,455 @@
...
+The above logic works in the opposite direction as well. If there is any change
+on the Bare Meta machine, BMO learns about the change(s) via ironic and a chain
+of events starts. Once BMO learns about the changes, it makes the required
+changes on the BMH.
+
+As discussed above, the management of Bare Metal Machines requires the
+interaction of multiple controllers via multiple CRs. However, the interaction
+is performed on a specified number of fields on each CR. i.e. A controller
+`watches` a specified number of fields in each CR and makes changes on a set of
+ fields. Before seeing the relationship, we need to see the CRs on their at two
```
Comment (maelk) [inline docs/architecture.md]: on their ?

File: docs/architecture.md [hunk 12]
```diff
@@ -0,0 +1,455 @@
...
+![components](images/components.png)
+
+As shown in the above Components' diagram, CAPBM is watching for changes in
+`Machine` CR and upon change, it makes changes on the `BareMetalMachine` CR.
+Similarly, it watches `Cluster` CR and makes changes on a related
+`BareMetalCluster` CR.
+
+The left most components, BMO controller and BareMetalHost(BMH) CR, are the
+closest to the Bare Metal Machine. If one wants to changes the state of a Bare
+Metal Machine, they modify the BMH CR. Upon change to the BMH, BMO interacts
```
Comment (maelk) [inline docs/architecture.md]: Calling a physical node a Bare Metal Machine can be very confusing considering that Bare Metal Machine is also one of our CRs

Comment (Xenwar) [reply docs/architecture.md]: Agree. 
Replaced by Bare Metal Servers, and some text added in the intro section for the terminology

**Comment by maelk** [comment]
@Xenwar You need to rebase this PR for the tests to pass

**Comment by Xenwar** [comment]
/ test-v1a3-integration

**Pull request merged**